File: go/dogma/angle.go
// This is free and unencumbered software released into the public domain.

package dogma

import "math"

// Angle represents an angle.
//
// See: https://dogma.dev/Angle
type Angle struct {
	radians float64
}

// AngleFromDegrees constructs an angle from degrees.
func AngleFromDegrees(degrees float64) Angle {
	return Angle{radians: degrees / 180.0 * math.Pi}
}

// AngleFromRadians constructs an angle from radians.
func AngleFromRadians(radians float64) Angle {
	return Angle{radians: radians}
}

// AngleFromTurns constructs an angle from turns.
func AngleFromTurns(turns float64) Angle {
	return Angle{radians: turns * 2 * math.Pi}
}

// Degrees returns the angle in degrees.
func (angle Angle) Degrees() float64 {
	return angle.radians / math.Pi * 180.0
}

// Radians returns the angle in radians.
func (angle Angle) Radians() float64 {
	return angle.radians
}

// Turns returns the angle in turns.
func (angle Angle) Turns() float64 {
	return angle.radians / (2 * math.Pi)
}

File: java/src/main/java/dogma/Angle.java
/* This is free and unencumbered software released into the public domain. */

package dogma;

import androidx.annotation.NonNull;

public interface Angle {
  /** Constructs an angle from degrees. */
  @NonNull
  public static Angle fromDegrees(final double degrees) {
    return new InRadians(degrees / 180.0 * Math.PI);
  }

  /** Constructs an angle from radians. */
  @NonNull
  public static Angle fromRadians(final double radians) {
    return new InRadians(radians);
  }

  /** Constructs an angle from turns. */
  @NonNull
  public static Angle fromTurns(final double turns) {
    return new InRadians(turns * 2 * Math.PI);
  }

  /** Returns this angle in degrees. */
  default public double getDegrees() {
    return this.getRadians() / Math.PI * 180.0;
  }

  /** Returns this angle in radians. */
  public double getRadians();

  /** Returns this angle in turns. */
  default public double getTurns() {
    return this.getRadians() / (2 * Math.PI);
  }

  class InRadians implements Angle {
    private static final long serialVersionUID = 1L;

    private final double radians;

    protected InRadians(final double radians) {
      this.radians = radians;
    }

    @Override
    public double getRadians() {
      return this.radians;
    }
  }
}
